src/test/java/org/jenkinsci/plugins/plaincredentials/BaseTest.java

```diff
@@ -38,74 +45,161 @@
-        assertThat(cred.getId(), is(CRED_ID));
-        // Update credential
-        store.updateCredentials(Domain.global(), cred, updatedCredential);
-        
-        // Look up all credentials again
-        credentials = CredentialsProvider.lookupCredentials(BaseStandardCredentials.class, r.jenkins, ACL.SYSTEM, Collections.<DomainRequirement>emptyList());
-   
-        // There is still 1 credential but the ID has been updated
-        assertThat(credentials.size(), is(1));
-        cred = credentials.get(0);
-        assertThat(cred, instanceOf(credential.getClass()));
-        assertThat(cred.getId(), is(UPDATED_CRED_ID));
-        
-        // Delete credential
-        store.removeCredentials(Domain.global(), cred);
-        
-        // Look up all credentials again
-        credentials = CredentialsProvider.lookupCredentials(BaseStandardCredentials.class, r.jenkins, ACL.SYSTEM, Collections.<DomainRequirement>emptyList());
-     
-        // There are no credentials anymore
-        assertThat(credentials.size(), is(0));
+    private <T extends BaseStandardCredentials> void testCreateUpdateDelete(T credential, T updatedCredential, Domain d) throws IOException {
+        boolean globalDomain = d == null;
+        Domain domainToUse = (globalDomain) ? Domain.global() : d;
+        CredentialsStore storeToUse = null;
+        SecurityContext ctx = null;
+        Authentication auth = null;
+
+        if (CredentialsScope.USER != credential.getScope()) {
+            storeToUse = CredentialsProvider.lookupStores(r.jenkins).iterator().next();;
+            auth = ACL.SYSTEM;
+        } else {
+            r.jenkins.setSecurityRealm(r.createDummySecurityRealm());
+            User alice = User.get("alice");
+            auth = alice.impersonate();
+            storeToUse = CredentialsProvider.lookupStores(alice).iterator().next();
+            ctx = ACL.impersonate(auth);
+        }
+
+        try {
+            // Add a credential
+            if (!globalDomain) {
+                storeToUse.addDomain(d);
+            }
+
+            storeToUse.addCredentials(domainToUse, credential);
+
+            // Look up all credentials
+            if (!globalDomain) {
+                DomainRequirement[] wrongIncludeDomainRequirements = {new HostnamePortRequirement("inclssude.com", 80)};
+                List<BaseStandardCredentials> credentialsWithWrongInclude = CredentialsProvider.lookupCredentials(BaseStandardCredentials.class, r.jenkins, auth, wrongIncludeDomainRequirements);
+                assertThat(credentialsWithWrongInclude.size(), is(0));
+
+                DomainRequirement[] excludeDomainRequirements = {new HostnamePortRequirement("exclude.com", 90)};
+                List<BaseStandardCredentials> credentialsWithExclude = CredentialsProvider.lookupCredentials(BaseStandardCredentials.class, r.jenkins, auth, excludeDomainRequirements);
+                assertThat(credentialsWithExclude.size(), is(0));
```

Comment:
Why not test that using the right domain requirements, you can find the credential?